generic error highlight generation

22a opened the pull request:
I've added a function that takes a source file and a line number (and optionally a number to indicate the number of surrounding lines to display above and below the highlighted line), and returns a pre-formatted DOM element.

It looks like this:
<img width="674" alt="screen shot 2017-03-25 at 03 49 02" src="https://cloud.githubusercontent.com/assets/7144173/24319257/2396c3d4-110e-11e7-88ef-68e1f6c0fc17.png">

To show it in use I've modified the `displayClashes` and `displayUnnamed` view functions to use it, those look like this:

<img width="749" alt="screen shot 2017-03-25 at 03 48 29" src="https://cloud.githubusercontent.com/assets/7144173/24319263/49241e44-110e-11e7-8500-e41e19f7a8a8.png">

I've done it this way to try and decouple the error highlighting logic from the bootstrap panels in order to make it easy to port to whatever UI we decide to go with because I don't like how it looks at the moment, much too bulky.

22a committed:
generic error highlight generation :massage:

22a commented:
ah ye, I've snapped a cucumber, will fix in the morning

c-brenn commented on panacea/web/static/js/view.js:
```diff
@@ -139,6 +156,45 @@ export const restoreFileForm = () => {
 };
 
+export const generateErrorHighlight = (code, errorLine, numContextLines = 2) => {
+  const lines = code.split('\n');
```
Do you think it makes sense to split the lines once we get the response? To avoid splitting multiple times.

c-brenn commented on panacea/web/static/js/app.js:
```diff
@@ -46,12 +46,15 @@ async function handleDrugsResponse(response) {
     }
 
+    // fetching the pml here, TODO: decide when is best to do this
+    const pml = await Request.pml(ast);
```
I've just realised there's a potential issue using `Request.pml(...)` here. The download button strips comments, so the line numbers may all be off.

This depends on how Yecc/Leex handle skipping tokens - I'll investigate a bit now. We may need to return the _exact_ PML they gave us.

We could alter Panacea to return the PML in the `/api/drugs` response or have JS read the file contents. Alternatively, we could change the parser to not strip comments - potentially messy.

c-brenn replied:
Seeing as the customer didn't buy any PML-TX features. The easiest approach might be to have JS read the file contents on submit.

We could then use that here, and remove the `/api/pml` endpoint entirely - the download button would be entirely JS as well.

Any thoughts?

c-brenn replied:
Same thing with whitespace - it all gets stripped. Submitting `no_drugs.pml` will highlight the wrong line as `unnamed` because there were blank lines in the source file that were stripped.

c-brenn commented on panacea/web/static/js/view.js:
```diff
@@ -139,6 +156,45 @@ export const restoreFileForm = () => {
   hideElement(loadingElement);
 };
 
+export const generateErrorHighlight = (code, errorLine, numContextLines = 2) => {
```
does this need to be exported?

c-brenn commented on panacea/web/static/js/view.js:
```diff
@@ -33,30 +33,47 @@ export const displayDrugs = drugs => {
-    const occurrences = sorted.map(occurrence => {
-      return `${occurrence.type} on line ${occurrence.line}`;
-    }).join(', ');
+    const wrapper = document.createElement('DIV');
```
Could this use `createElementWithClass(...)`?

c-brenn reviewed (changes requested):
this is beautiful - excellent work.

Some minor bits to discuss.

c-brenn committed:
ensure atoms are converted to strings

22a committed:
retrieve file locally, split once, clean exports

22a commented:
I've changed it so that the pml file is pulled from the file input, should correct any line number errors caused by the PML-TX parsing stripping comments / whitespace.

This brought up the discussion that it may be worthwhile decommissioning the AST->PML endpoint entirely in favour of triggering the file download with the original file provided by the user. We had some dirty code to do that a few pr's ago: https://github.com/tom-and-the-toothfairies/pathways/blob/6dbda89fe1e3c1507770d37d22c3a3ad41a1077e/panacea/web/static/js/app.js#L122-L130 which would work now that we have the raw file contents again.

c-brenn commented on panacea/web/static/js/app.js:
```diff
@@ -91,3 +95,19 @@ async function handleDdisResponse(response, urisToLabels) {
+
+const getFileContents = callback => {
+  const file = View.fileInputElement.files[0];
+  const fileName = file.name;
```
Doesn't look like `fileName` is used anywhere. Any reason to keep it?

c-brenn commented on panacea/web/static/js/view.js:
```diff
@@ -139,10 +156,47 @@ export const restoreFileForm = () => {
+
+  const codeBlockElement = createElementWithClass('PRE','code-block');
+
+  displayLines.forEach((lineContents, currentIndex) => {
```
Not too familiar with JS idioms, but in other places we're using:

```javascript
for (const x of xs)
```

Whereas here we use `.forEach(...)`. Any particular reason? If not, I'd go for consistency everywhere.

c-brenn replied:
o i c, the index

c-brenn reviewed (changes requested):
very nice indeed

c-brenn committed:
refactors some javascript [skip ci]

* adds new module for code block generation
* adds a Util module for shared view code

houli commented on panacea/web/static/css/app.css:
```diff
@@ -38,3 +38,42 @@
+  font-size: .8em;
+  line-height: 1.4;
+  white-space: normal;
+  font-family: menlo, consolas, monospace;
```
Add `Ubuntu Mono` here after Consolas?

houli commented on panacea/web/static/js/code-block.js:
```diff
@@ -0,0 +1,41 @@
+
+  const displayLines = lines.slice(startIndex, endIndex);
+
+  const codeBlock = Util.createElementWithClass('PRE', 'code-block');
```
Don't think we need the `CAPS` on all of these

houli commented on panacea/web/static/js/code-block.js:
```diff
@@ -0,0 +1,41 @@
+
+  displayLines.forEach((lineContents, index) => {
+    const lineNumber = startIndex + index + 1;
+    const isHighlighted = lineNumber == highlightLineNumber;
```
`===`

c-brenn replied:
*bleugh*

houli commented on panacea/web/static/js/code-block.js:
```diff
@@ -0,0 +1,41 @@
+  })
+
+  return codeBlock;
+}
```
`;`

houli commented on panacea/web/static/js/code-block.js:
```diff
@@ -0,0 +1,41 @@
+  lineContainer.appendChild(lineContentSpan);
+
+  return lineContainer;
+}
```
`;`

houli commented on panacea/web/static/js/util.js:
```diff
@@ -0,0 +1,6 @@
+  const element = document.createElement(elementName);
+  element.classList.add(className);
+  return element;
+}
```
`;`

c-brenn committed:
:wow: lots of cleaning

c-brenn committed:
refactor warning display

c-brenn committed:
adds new warnings to the change log

c-brenn committed:
moves `off by 1` fix to where it's actually needed

The `+ 1` is added because `Array.Protptype.splice()` splices up-to but not
including the end index.

This moves the `+ 1` into the splice call to avoid messing up `endIndex` in
other places.
